### ISSUE TITLE
Regression: @Spy with @InjectMock are not injected anymore

### ISSUE THREAD

**Issue by arnor2000** [mockito/mockito#1631]
Hi,

 - [x] The mockito message in the stacktrace have useful information, but it didn't help
    java.lang.NullPointerException
	at org.test.TestedService.servivce(TestedService.java:16)
	at org.test.TestedServiceTest.mockSpyAndInjectAll(TestedServiceTest.java:33)
 - [x] The problematic code (if that's possible) is copied here;
    Fix of #1587 (Remove unnecessary loop from InjectingAnnotationEngine) has introduce a regression.
 - [x] Provide versions (mockito / jdk / os / any other relevant information)
    Issue can be reproduce starting from Mockito version 2.23.15.
 - [x] Provide a [Short, Self Contained, Correct (Compilable), Example](http://sscce.org) of the issue
    Please find attached [inject-spy.zip](https://github.com/mockito/mockito/files/2905749/inject-spy.zip).
    execute `./gradlew check`
- [x] Read the [contributing guide](https://github.com/mockito/mockito/blob/master/.github/CONTRIBUTING.md)

We are experiencing a regression behavior caused by fix of #1587 (Remove unnecessary loop from InjectingAnnotationEngine) which prevent us to update from Mockito 2.23.4 to newer versions.

```java
public class TestedServiceTest extends AbstractTest {

    @InjectMocks
    private TestedService testedService;

    @InjectMocks
    @Spy
    private SpiedCompositionService spiedCompositionService;

    @Mock
    private MockedDao mockedDao;
    ...
}
```

As you can see in this snippet and example project attached, with Mockito 2.23.14 when we try to inject a mock in a spy and this spy in the tested object, it works as expected and everything is injected. But when we switch to Mockito 2.23.15, this behavior doesn't work anymore, the spy is not injected to tested object and obviously causes a NullPointerException.

We also figured out injection of injected spies was working only because test class extends has an abstract test class.

I understand this test case is not a state of the art way to do unit test, our use case is more a component test because several classes work together to do a coherent unit of work and it makes more sense to test them together.

We are relying on the behavior provided up to Mockito 2.23.14 and the only workaround I see is to expose attributes to package visibility and do injection by hand.

The loop removed by #1587 was injecting injected spies newly instantiated by first loop. The depth of injection was relying on the depth of test hierarchy.

**Comment by jpasquers** [comment]
Is there any update on this? This has broken our tests as well attempting to upgrade our version of Mockito. The usage of an @InjectMocks as well as a @Spy & @InjectMocks makes sense from the perspective of integration testing of two components that act as one cohesive unit.

**Comment by tokuhirom** [comment]
Here's a pseudo code to describe the issues around this issue.

# Before 2.23.15:	

```
while (clazz != Object.clazz) {
	val mocks = aggregateSpyOrMockFields(clazz.getMockFields());
  val fields = aggregateInjectMocksFields(clazz);
	val ret = tryConstructorInjection(clazz, fields, mock);
  if (! ret.fieldWasInitializedUsingConstructorInjection()) {
    trySetterOrPropertyInjection(clazz, fields, mock);
  }
	clazz = clazz.getSuperclass();
}
```
# After 2.23.15:
```
	val mocks = aggregateSpyOrMockFields(clazz.getMockFields());
  val fields = aggregateInjectMocksFields(clazz);
	val ret = tryConstructorInjection(clazz, fields, mock);
  if (! ret.fieldWasInitializedUsingConstructorInjection()) {
    trySetterOrPropertyInjection(clazz, fields, mock);
  }
```

# Details

Before 2.23.15, if the `@InjectMocks` field has a field that not listed in the constructor argument will fill by setter/property injection at the 2nd loop 😃
It may not the expected behaviors by Mockito authors, but some users' code depends on this behavior.

My suggestion is to revert the [Fixes #1587 : Remove unnecessary loop from InjectingAnnotationEngine by LihMeh · Pull Request #1588 · mockito/mockito · GitHub](https://github.com/mockito/mockito/pull/1588) at this time.

**Comment by TimvdLippe** [comment]
Feel free to open a PR with regression test and I will be happy to merge it. Thanks!

**Comment by TimvdLippe** [comment]
This should be fixed with #1810
